libcaf_net/caf/net/multiplexer.hpp

```diff
@@ -60,179 +55,101 @@ class CAF_NET_EXPORT multiplexer : public detail::atomic_ref_counted,
+  // virtual void run_actions() = 0;
 
   /// Sets the thread ID to `std::this_thread::id()`.
-  void set_thread_id();
+  virtual void set_thread_id() = 0;
 
   /// Runs the multiplexer until no socket event handler remains active.
-  void run();
-
-protected:
-  // -- utility functions ------------------------------------------------------
-
-  /// Handles an I/O event on given manager.
-  void handle(const socket_manager_ptr& mgr, short events, short revents);
-
-  /// Returns a change entry for the socket at given index. Lazily creates a new
-  /// entry before returning if necessary.
-  poll_update& update_for(ptrdiff_t index);
-
-  /// Returns a change entry for the socket of the manager.
-  poll_update& update_for(socket_manager* mgr);
-
-  /// Writes `opcode` and pointer to `mgr` the the pipe for handling an event
-  /// later via the pollset updater.
-  /// @warning assumes ownership of @p ptr.
-  template <class T>
-  void write_to_pipe(uint8_t opcode, T* ptr);
-
-  /// @copydoc write_to_pipe
-  template <class Enum, class T>
-  std::enable_if_t<std::is_enum_v<Enum>> write_to_pipe(Enum opcode, T* ptr) {
-    write_to_pipe(static_cast<uint8_t>(opcode), ptr);
-  }
-
-  /// Queries the currently active event bitmask for `mgr`.
-  short active_mask_of(const socket_manager* mgr) const noexcept;
-
-  // -- member variables -------------------------------------------------------
-
-  /// Bookkeeping data for managed sockets.
-  pollfd_list pollset_;
-
-  /// Maps sockets to their owning managers by storing the managers in the same
-  /// order as their sockets appear in `pollset_`.
-  manager_list managers_;
-
-  /// Caches changes to the events mask of managed sockets until they can safely
-  /// take place.
-  poll_update_map updates_;
-
-  /// Stores the ID of the thread this multiplexer is running in. Set when
-  /// calling `init()`.
-  std::thread::id tid_;
-
-  /// Guards `write_handle_`.
-  std::mutex write_lock_;
-
-  /// Used for pushing updates to the multiplexer's thread.
-  pipe_socket write_handle_;
-
-  /// Points to the owning middleman.
-  middleman* owner_;
-
-  /// Signals whether shutdown has been requested.
-  bool shutting_down_ = false;
-
-  /// Pending actions via `schedule`.
-  std::deque<action> pending_actions_;
-
-  /// Keeps track of watched disposables.
-  std::vector<disposable> watched_;
+  virtual void run() = 0;
 
 private:
-  // -- constructors, destructors, and assignment operators --------------------
+  // -- internal getter for the pollset updater --------------------------------
 
-  explicit multiplexer(middleman* parent);
+  virtual std::deque<action>& pending_actions() = 0;
 
   // -- internal callbacks the pollset updater ---------------------------------
 
-  void do_shutdown();
-
-  void do_register_reading(const socket_manager_ptr& mgr);
+  virtual void do_shutdown() = 0;
 
-  void do_start(const socket_manager_ptr& mgr);
+  virtual void do_start(const socket_manager_ptr& mgr) = 0;
```

Comment:
This is forcing the implementation into actually having a `deque<action>`. Please remove.

The pollset updater is an implementation detail of the default implementation. Consequently, the pollset updater can hold a pointer to the actual implementation type.

Also, the pollset updater and the default multiplexer are both implementation details that really shouldn't be part of the public API. Please move them to `multiplexer.cpp`, they don't need to have their own headers. Both of them might eventually become obsolete (cf. #1424).